File: fastLPR_R/R/fastlpr_dof.R
# Copyright (c) 2024-2025 Ying Wang, Min Li
# SPDX-License-Identifier: GPL-3.0-or-later

#' Degree-of-freedom calculations
#'
#' Mirrors: fastLPR/utility/core/fastLPR_dof.m
#'
#' @keywords internal
fastlpr_dof <- function(regs, weights_list, sum_weights_list) {
  Tx <- regs$Tx
  dh <- length(weights_list)

  df <- numeric(dh)
  diagS_mat <- matrix(0, nrow = Tx, ncol = dh)
  pdof_vec <- numeric(dh)

  for (j in seq_len(dh)) {
    weights <- weights_list[[j]]
    row_sum <- sum_weights_list[[j]]
    diagS <- diag(weights) / row_sum
    df[j] <- sum(diagS)
    diagS_mat[, j] <- diagS
    pdof_vec[j] <- mean(1 - diagS)
  }

  regs$df_m <- df
  regs$dof <- df[which.min(regs$gcv_yhat$gcv_m)]
  regs$diagS <- diagS_mat
  regs$pdof_m <- pdof_vec
  regs
}

#' Estimate degrees of freedom using Hutchinson's randomized trace estimator
#'
#' This is MUCH faster than the naive O(N^2) approach because:
#' 1. No need to compute the full NxN weight matrix
#' 2. Uses existing fast NUFFT-based regression (O(N + M log M))
#' 3. Only 5-10 random samples needed for stable estimate
#'
#' COMPLEXITY:
#'   Naive approach: O(N^2 * dh)  [NxN matrix for each bandwidth]
#'   Hutchinson:     O((N + M log M) * dh * num_samples)  [just M=100-500 typically]
#'   Speedup:        50-100x for large N (N=5000+)
#'
#' Algorithm (from MATLAB fastLPR_dof.m):
#' 1. Generate random vectors: p ~ N(0,I)
#' 2. Compute fitted values: p_hat = H*p (using fast method!)
#' 3. Estimate trace: tr(I-H) ?(1/N) * p'*(p - p_hat) / (p'*p)
#' 4. Repeat for multiple samples and compute statistics
#'
#' VECTORIZATION (2025-11-21):
#' - Process ALL samples simultaneously (no loop!)
#' - Treats num_samples as multi-response dimension
#' - 3-5x speedup vs sequential processing
#'
#' Mirrors: fastLPR/utility/core/fastLPR_dof.m
#'
#' @param regs Regression structure from fastlpr_s
#' @param num_samples Number of random samples (default: 10)
#'
#' @return List with DOF estimates for each bandwidth
#' @keywords internal
fastlpr_dof_hutchinson <- function(regs, num_samples = 10) {

  Tx <- regs$Tx
  dh <- regs$dh
  dy <- regs$dy

  # ==========================================================================
  # FIX 2025-12-01: Support external random vectors for exact reproducibility
  # If regs$opt$dof_random_vectors is provided (e.g., from MATLAB reference file),
  # use those exact vectors instead of generating new ones.
  # This enables exact matching between R and MATLAB DOF computations.
  # ==========================================================================
  if (!is.null(regs$opt$dof_random_vectors)) {
    p <- regs$opt$dof_random_vectors
    # Validate dimensions
    if (nrow(p) != Tx || ncol(p) != num_samples) {
      stop(sprintf(
        "dof_random_vectors must be (%d x %d), got (%d x %d)",
        Tx, num_samples, nrow(p), ncol(p)
      ))
    }
  } else {
    # Fallback: Generate random probe vectors for DOF estimation
    # NOTE: R's rnorm will NOT match MATLAB's randn because:
    #   - R uses inverse-CDF method, MATLAB uses Ziggurat algorithm
    #   - Even RNGkind("Mersenne-Twister") doesn't help (different seed init)
    # For cross-language reproducibility, use opt$dof_random_vectors from MATLAB.
    # For R-only use, statistical equivalence is sufficient (same N(0,1) distribution).
    set.seed(42)  # Fixed seed for R-internal reproducibility
    p <- matrix(rnorm(Tx * num_samples), nrow = Tx, ncol = num_samples)
  }

  # ===== VECTORIZED SAMPLE PROCESSING =====
  # Process ALL samples at once by treating them as multi-response data
  # This matches MATLAB's approach: mq = fastLPR_reg(regs, p)
  #
  # Key insight: fastlpr_reg interprets multi-column y as multiple responses (dy dimension)
  # Passing p with shape (Tx, num_samples) processes ALL samples in ONE call!
  # Output: mq with shape (Ng, dh, num_samples) - exactly what we need
  #
  # OLD APPROACH (SLOW): Loop over num_samples, call fastlpr_reg separately
  # NEW APPROACH (FAST): Single call, broadcast across samples automatically

  # CRITICAL: Use the ORIGINAL regs$s (design matrix) that was already computed!
  # The design matrix depends on kernel weights at data locations, NOT on response values.
  # Recomputing s for random vectors p was the bug causing 47x underestimation.
  #
  # MATLAB (correct): mq = fastLPR_reg(regs, p)  # Uses precomputed regs.s
  # R (was wrong): regs_temp$s <- NULL; regs_temp <- fastlpr_s(regs_temp)  # Recomputed s
  # R (now correct): Just call fastlpr_reg with original regs and new response p
  #
  # FIX 2025-12-04: Temporarily set y_isreal=TRUE for DOF computation
  # The random probe vectors p are ALWAYS real, so the output should be real.
  # But regs$y_isreal might be FALSE if original data y was complex.
  # This prevents unnecessary complex computations and allows Rcpp interpolation.
  y_isreal_orig <- regs$y_isreal
  regs$y_isreal <- TRUE  # Force real output for real probe vectors
  mq <- fastlpr_reg(regs, p)
  regs$y_isreal <- y_isreal_orig  # Restore original setting
  
  # Ensure output is real (discard any numerical noise from FFT)
  if (is.complex(mq)) {
    mq <- Re(mq)
  }


  # MATLAB: mq = reshape(mq, [regs.N, regs.dh*regs.opt.num_dof_sample])
  # Reshape mq to 2D: (Ng, dh*num_samples)
  # This allows interpolation to treat each column as a separate response
  mq_dims <- dim(mq)
  if (is.null(mq_dims)) {
    # Vector: reshape to (Ng, dh*num_samples)
    Ng <- length(mq)
    mq <- matrix(mq, nrow = Ng, ncol = dh * num_samples)
  } else if (length(mq_dims) == 1) {
    # 1D array: reshape to (Ng, dh*num_samples)
    Ng <- mq_dims[1]
    mq <- matrix(mq, nrow = Ng, ncol = dh * num_samples)
  } else if (length(mq_dims) == 2) {
    # Already 2D, but might need reshaping if not (Ng, dh*num_samples)
    # Keep as is if dimensions match, otherwise reshape
    if (ncol(mq) != dh * num_samples) {
      Ng <- nrow(mq)
      mq <- matrix(mq, nrow = Ng, ncol = dh * num_samples)
    }
  } else {
    # 3D or higher: flatten to 2D
    # OPTIMIZATION v5 (2025-12-04): Use aperm + matrix instead of loop
    # R's fastlpr_reg returns 3D array (Ng, dh, num_samples)
    # MATLAB returns 2D (flattened) with columns ordered as (dh1, dh2, ..., dhN, sample1, sample2, ...)
    # We need to convert (Ng, dh, num_samples) -> (Ng, dh*num_samples)
    Ng <- mq_dims[1]
    dh_actual <- mq_dims[2]
    num_samples_actual <- mq_dims[3]

    # FAST: aperm reorders dimensions, then matrix flattens last two dims
    # aperm(mq, c(1, 2, 3)) keeps same order: (Ng, dh, num_samples)
    # matrix(..., nrow=Ng) flattens (dh, num_samples) -> (dh*num_samples) in column-major order
    # This gives columns: [dh1,s1], [dh2,s1], ..., [dhN,s1], [dh1,s2], [dh2,s2], ...
    mq <- matrix(mq, nrow = Ng, ncol = dh_actual * num_samples_actual)
  }

  # Now mq is 2D: (Ng, dh*num_samples)

  # ===== END VECTORIZED SECTION =====

  # ===== INTERPOLATION: RCPP vs PURE R =====
  # MATLAB uses griddedInterpolant which handles all columns at once
  # R's approxfun requires looping (slow!)
  # Solution: Use Rcpp batch interpolation if available

  use_rcpp <- FALSE
  if (!is.null(regs$opt$use_rcpp)) {
    use_rcpp <- regs$opt$use_rcpp && requireNamespace("Rcpp", quietly = TRUE)
  } else {
    # Auto-detect: Use Rcpp if available and many bandwidths
    # FIX 2025-12-14: Use get0() instead of exists() because exists() doesn't
    # reliably find functions in the package namespace when called from within
    dof_interp_batch_fn <- tryCatch({
      fn <- get0("dof_interpolate_batch", envir = globalenv(), inherits = FALSE)
      if (is.null(fn)) fn <- get0("dof_interpolate_batch", envir = asNamespace("fastlpr"), inherits = FALSE)
      fn
    }, error = function(e) NULL)
    use_rcpp <- (dh * num_samples > 50) && is.function(dof_interp_batch_fn)
  }

  if (regs$dx == 1) {
    # 1D case
    # MATLAB approach:
    # 1. mq is 2D: (Ng, dh*num_samples)
    # 2. Interpolate each column separately
    # 3. phat_flat is 2D: (Tx, dh*num_samples)
    # 4. Reshape phat_flat to 3D: (Tx, dh, num_samples)
    # 5. Permute to (Tx, num_samples, dh)

    grid_raw <- as.numeric(regs$xlist[[1]])  # Grid points in RAW space
    x_raw <- as.numeric(regs$xraw[, 1])      # Data points in RAW space

    # OPTIMIZATION: Use unified rcpp_interp_batch_nd (N-D, OpenMP parallelized)
    # mq: (Ng, dh * num_samples), output: (Tx, dh * num_samples)
    dims <- as.integer(c(length(grid_raw), ncol(mq)))
    phat_flat <- rcpp_interp_batch_nd(
      list(as.numeric(grid_raw)),
      as.numeric(mq),
      dims,
      matrix(x_raw, ncol = 1)
    )

    # MATLAB: phat = reshape(phat, [regs.Ty, regs.dh, regs.opt.num_dof_sample])
    # Reshape to 3D: (Tx, dh, num_samples)
    phat <- array(phat_flat, dim = c(Tx, dh, num_samples))
    
    # MATLAB: phat = permute(phat, [1,3,2])
    # Permute to (Tx, num_samples, dh)
    phat <- aperm(phat, c(1, 3, 2))  # aperm is R's permute
    
  } else {
    # Multi-dimensional case - more complex interpolation needed
    stop("Hutchinson DOF estimator not yet implemented for dx > 1")
  }

  # Compute pdof using Hutchinson's estimator
  # MATLAB:
  #   phat: (Ty, num_samples, dh) after permute
  #   p: (Ty, num_samples)
  #   pdof = sum(p.*(p-phat)) ./ sum(p.*p)  # Broadcasting and sum along dim 1
  #   pdof: (1, num_samples, dh)
  #   pdof = permute(pdof, [2,3,1]) -> (num_samples, dh, 1)
  
  # R implementation (OPTIMIZED):
  # phat: (Tx, num_samples, dh)
  # p: (Tx, num_samples)
  
  # OPTIMIZATION: Use array recycling instead of explicit loop
  # R automatically recycles p along the 3rd dimension when dimensions match
  p_broadcast <- array(p, dim = c(Tx, num_samples, dh))
  
  # OPTIMIZATION: Use colSums instead of apply(..., sum) - 15x faster
  # For 3D array, we need to collapse first dimension
  # Reshape to 2D: (Tx, num_samples*dh), apply colSums, then reshape back
  diff_array <- p_broadcast * (p_broadcast - phat)
  dim_diff <- dim(diff_array)
  diff_2d <- matrix(diff_array, nrow = dim_diff[1], ncol = dim_diff[2] * dim_diff[3])
  numerator_flat <- colSums(diff_2d)
  numerator <- matrix(numerator_flat, nrow = num_samples, ncol = dh)
  
  # OPTIMIZATION: colSums for denominator
  denominator <- colSums(p * p)  # (num_samples,)
  
  # Broadcast denominator to (num_samples, dh) for division
  denominator_broadcast <- matrix(rep(denominator, dh), nrow = num_samples, ncol = dh)

  # Compute pdof: (num_samples, dh)
  pdof <- numerator / denominator_broadcast

  # Compute df: df = Tx * (1 - pdof)
  df <- Tx * (1 - pdof)


  # Clamp pdof to valid range [0, 1]
  # Hutchinson estimator has variance and can give values outside [0,1]
  # NOTE: pmax/pmin destroy matrix dimensions, so we need to restore them
  pdof_dims <- dim(pdof)
  pdof <- pmax(0, pmin(1, pdof))
  dim(pdof) <- pdof_dims  # Restore matrix dimensions

  # FIX 2025-12-01: Complex-valued responses need 2x DOF adjustment
  # MATLAB code:
  #   if ~regs.y_isreal
  #       pdof(:,:,~regs.y_isreal)=2*pdof(:,:,~regs.y_isreal)-1;
  #   end
  # This is because complex-valued data has 2 degrees of freedom per observation
  if (!is.null(regs$y_isreal) && !all(regs$y_isreal)) {
    # Apply 2x DOF adjustment for complex responses
    pdof <- 2 * pdof - 1
    # Re-clamp after adjustment
    pdof <- pmax(0, pmin(1, pdof))
    dim(pdof) <- pdof_dims
  }

  # CRITICAL FIX (2025-11-21): Compute df from CLAMPED pdof
  # When pdof > 1.0, this caused negative df values - compute df AFTER clamping
  df <- Tx * (1 - pdof)

  # Ensure 2D matrices for colMeans/apply (handles dh=1 case)
  pdof <- as.matrix(pdof)
  df <- as.matrix(df)

  # Compute statistics across samples
  pdof_m <- colMeans(pdof)
  pdof_sd <- apply(pdof, 2, sd)
  df_m <- colMeans(df)
  df_sd <- apply(df, 2, sd)

  # ==========================================================================
  # Compute GCV penalty and its standard deviation
  # CRITICAL: Must compute penalty samples FIRST, then take mean/sd
  # The delta method approximation was causing incorrect gcv_sd values
  # ==========================================================================

  # Clamp pdof samples for penalty computation (prevents division by zero)
  # pdof is (num_samples, dh) matrix
  pdof_for_penalty <- pmax(pdof, 0.01)  # Minimum 1% residual DOF

  # Compute penalty samples: penalty = 1 / pdof^2
  penalty_samples <- 1 / (pdof_for_penalty^2)  # (num_samples, dh) matrix

  # Compute mean and std of penalty (matches MATLAB exactly)
  pdof_inv_m <- colMeans(penalty_samples)  # Mean penalty for each bandwidth
  pdof_inv_sd <- apply(penalty_samples, 2, sd)  # Std of penalty for each bandwidth

  return(list(
    pdof_m = pdof_m,
    pdof_sd = pdof_sd,
    df_m = df_m,
    df_sd = df_sd,
    pdof_inv_m = pdof_inv_m,
    pdof_inv_sd = pdof_inv_sd,
    num_samples = num_samples
  ))
}

File: fastLPR/utility/core/fastlpr_dof.m
function [regs]=fastlpr_dof(regs)
% FASTLPR_DOF - Estimate degrees of freedom using Hutchinson's method
%
% Estimates the effective degrees of freedom (DOF) for each bandwidth using
% a randomized trace estimator (Hutchinson's method). The DOF is used for
% GCV bandwidth selection and variance estimation.
%
% Syntax:
%   regs = fastlpr_dof(regs)
%
% Input:
%   regs - Regression structure from fastlpr_s containing:
%          .opt.calc_dof - Whether to compute DOF (boolean)
%          .opt.num_dof_sample - Number of random samples (default: 10)
%          .Ty - Number of data points
%          .dh - Number of bandwidths
%          .dy - Number of responses
%          .y_isreal - Logical array indicating real-valued responses
%
% Output:
%   regs - Updated structure with DOF statistics:
%          .pdof_m - Mean proportion of residual DOF: (1/N)tr(I-H)
%          .pdof_sd - Std of pdof
%          .df_m - Mean degrees of freedom: tr(H)
%          .df_sd - Std of df
%          .pdof_inv_m - Mean GCV penalty: 1/pdof
%          .pdof_inv_sd - Std of GCV penalty
%
% Algorithm (Hutchinson's Method):
%   1. Generate random vectors: p ~ N(0,I)
%   2. Compute fitted values: p_hat = H*p
%   3. Estimate trace: tr(I-H) ~= (1/N) * p'*(p - p_hat) / (p'*p)
%   4. Repeat for multiple samples and compute statistics
%   5. Clamp pdof to [0,1] for numerical stability
%   6. Compute GCV penalty with additional clamping
%
% GCV Penalty Clamping:
%   - pdof estimates (1/N)tr(I-H), theoretically in [0,1]
%   - For GCV penalty 1/pdof, clamp pdof to minimum 0.01
%   - This prevents penalty explosion when pdof -> 0 (overfitting)
%   - Maximum penalty = 1/0.01 = 10,000 (reasonable upper bound)
%
% Complex-Valued Responses:
%   - Complex responses have 2x DOF (real and imaginary parts)
%   - Adjustment: pdof_complex = 2*pdof_real - 1
%
% References:
%   - Hutchinson (1990): A stochastic estimator of the trace of the
%     influence matrix for Laplacian smoothing splines
%   - Girard (1989): A fast 'Monte-Carlo cross-validation' procedure
%     for large least squares problems with noisy data
%
% Author: Ying Wang, Min Li
% Create Time: 2021
% Copyright (c): 2020-2025 Ying Wang, yingwangrigel@gmail.com,
%                Min Li, minli.231314@gmail.com
% Joint China-Cuba LAB, UESTC
% License: GNU General Public License v3.0 (see LICENSE file)

% Formula: (1/n)w'*(w-what) == (1/n)tr(I-A)
if regs.opt.calc_dof
    % Use external random vectors if provided (for exact cross-language matching)
    % Otherwise use dof_seed for reproducible random generation
    if isfield(regs.opt, 'dof_random_vectors') && ~isempty(regs.opt.dof_random_vectors)
        p = regs.opt.dof_random_vectors;
        % Validate dimensions
        if size(p, 1) ~= regs.Ty || size(p, 2) ~= regs.opt.num_dof_sample
            error('fastLPR:dof:invalidVectors', ...
                'dof_random_vectors must be (%d x %d), got (%d x %d)', ...
                regs.Ty, regs.opt.num_dof_sample, size(p, 1), size(p, 2));
        end
    else
        % Use dof_seed for reproducible DOF estimation (like Python's dof_seed)
        % Save current RNG state, set seed, generate vectors, restore state
        rng_state = rng;
        rng(regs.opt.dof_seed);
        p = randn(regs.Ty, regs.opt.num_dof_sample);
        rng(rng_state);  % Restore original RNG state
    end
    % Store the vectors used (for saving to reference files)
    regs.dof_random_vectors = p;
    
    mq=fastlpr_reg(regs,p);%N*ones(1,regs.dx) regs.dh regs.opt.num_dof_sample
    
    mq=reshape(mq,[regs.N, regs.dh*regs.opt.num_dof_sample]);
%     method ='griddedInterpolant';
%     opt.Method='spline';
%     opt.ExtrapolationMethod ='linear';
%     [~,phat]=fastlpr_gridinterp(regs,mq,[],method,opt);
    [~,phat]=fastlpr_gridinterp(regs,mq);
    phat=real(phat);
    phat=reshape(phat,[regs.Ty,regs.dh,regs.opt.num_dof_sample]);
    phat=permute(phat,[1,3,2]);
    
    pdof=(sum(p.*(p-phat))./sum(p.*p));
    pdof=permute(pdof,[2,3,1]);
    pdof=repmat(pdof,1,1,regs.dy);
    if ~regs.y_isreal
        pdof(:,:,~regs.y_isreal)=2*pdof(:,:,~regs.y_isreal)-1; % complex value has 2*df
    end

    % Clamp pdof to valid range [0, 1]
    % pdof estimates (1/N)tr(I-H), which theoretically should be in [0, 1]
    % However, Hutchinson estimator has variance and can give pdof > 1 or pdof < 0
    % This happens especially with very small bandwidths (high DOF, tr(I-H)~0)
    % Clamping prevents negative df and improves numerical stability
    pdof_clamped = max(0, min(1, pdof));  % Clamp to [0, 1]

    % Check if clamping was needed (for diagnostic purposes)
    % Count violations per bandwidth (average across samples)
    pdof_mean = mean(pdof, 1);  % Average across samples
    n_bandwidths_violated = sum(any(pdof_mean < 0 | pdof_mean > 1, 3));

    % Only warn if violations are significant (not just numerical noise)
    if any(pdof_mean(:) < -0.05 | pdof_mean(:) > 1.05)
        max_violation = max(abs([min(pdof_mean(:)), max(pdof_mean(:))-1]));
        warning('fastLPR:dof:outOfRange', ...
            ['pdof out of valid range [0,1] for %d bandwidth(s) (max violation: %.2f).\n' ...
             'This is normal with extreme bandwidths due to Hutchinson estimator variance.\n' ...
             'Values are automatically clamped. Consider using a narrower bandwidth range.'], ...
            n_bandwidths_violated, max_violation);
    end

    df=regs.Ty-regs.Ty*pdof_clamped;

    % Compute GCV penalty with clamping for numerical stability
    %
    % The GCV formula is: GCV = MSE / pdof^2
    % where pdof = (1/N)tr(I-H) is the proportion of residual DOF
    %
    % Problem: When pdof -> 0 (small bandwidth, overfitting):
    % 1. The penalty 1/pdof^2 -> inf (explodes)
    % 2. The variance of 1/pdof^2 -> inf (unstable for 1-SE rule)
    %
    % Solution: Clamp pdof to a reasonable minimum
    % - Use pdof_min = 0.01 (1% residual DOF)
    % - This means at least N/100 residual DOF
    % - For N=1000: at least 10 residual DOF
    % - For N=73508 (qEEG): at least 735 residual DOF
    % - Bandwidths with pdof < 0.01 are considered equally bad (overfitting)
    %
    % This gives maximum penalty = 1/0.01^2 = 10,000 (reasonable upper bound)
    pdof_for_penalty = max(pdof_clamped, 0.01);  % Clamp to minimum 1% residual DOF

    % Compute penalty and its statistics
    penalty = 1 ./ pdof_for_penalty.^2;  % GCV penalty: 1/pdof^2
    regs.pdof_inv_m = permute(mean(penalty, 1), [2,3,1]);
    regs.pdof_inv_sd = permute(std(penalty, [], 1), [2,3,1]);

    % For backward compatibility, also store log-space versions
    % (not used in GCV calculation, but kept for diagnostics)
    regs.log_gcv_penalty_m = permute(mean(log(penalty), 1), [2,3,1]);
    regs.log_gcv_penalty_sd = permute(std(log(penalty), [], 1), [2,3,1]);

    % Store other DOF statistics
    regs.df_m = permute(mean(df, 1), [2,3,1]);
    regs.df_sd = permute(std(df, [], 1), [2,3,1]);
    regs.pdof_m = permute(mean(pdof_clamped, 1), [2,3,1]);  % Use clamped pdof
    regs.pdof_sd = permute(std(pdof_clamped, [], 1), [2,3,1]);  % Use clamped pdof
else
    regs.log_gcv_penalty_m = [];
    regs.log_gcv_penalty_sd = [];
    regs.pdof_inv_sd = [];
    regs.pdof_inv_m = [];
    regs.df_m = [];
    regs.df_sd = [];
    regs.pdof_m = [];
    regs.pdof_sd = [];
end

end
